R/github-pages.R

```diff
@@ -0,0 +1,164 @@
+#' Configure a GitHub Pages site
+#'
+#' Activates or reconfigures a GitHub Pages site for a project hosted on GitHub.
+#' This function anticipates two specific usage modes:
+#' * Publish from the root directory of a `gh-pages` branch, which is assumed to
+#'   be only (or at least primarily) a remote branch.
+#'   - Typically the `gh-pages` branch is managed by an automatic "build and
+#'     deploy" job, such as the one configured by
+#'     [`use_github_action("pkgdown")`][use_github_action()].
+#' * Publish from the `"/docs"` directory of a "regular" branch, probably the
+#'   repo's default branch.
+#'   - The user is assumed to have a plan for how they will manage the content
+#'     below `"/docs"`.
+#'
+
+#' @param branch Branch for the site source. Default is `gh-pages`, coupled with
+#'   `path = "/"`. These are the usethis defaults, but more importantly, they
+#'   are GitHub's defaults, i.e. when a `gh-pages` branch is first created, it
+#'   is *automatically* published to Pages, using the source found in `"/"`. If
+#'   a `gh-pages` branch does not yet exist on the host, `use_github_pages()`
+#'   creates an empty, orphan remote branch.
+#'
+#'   The most common alternative is the repo's default branch, coupled with
+#'   `path = "/docs"`. It is the user's responsibility to ensure that this
+#'   `branch` pre-exists on the host.
+#' @param path Path to the site source. Note that GitHub only supports `"/"` or
+#'   `"/docs".`
+
+#' @param cname Optional, custom domain name. The `NA` default means "don't set
+#'   or change this", whereas a value of `NULL` removes any previously
+#'   configured custom domain.
+#'
+
+#' @seealso
+#' * <https://docs.github.com/en/free-pro-team@latest/github/working-with-github-pages>
+#' * <https://docs.github.com/en/free-pro-team@latest/rest/reference/repos#pages>
+
+#' @return Nothing
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' use_github_pages()
+#' }
+use_github_pages <- function(branch = "gh-pages", path = "/", cname = NA) {
+  stopifnot(is_string(branch), is_string(path))
+  stopifnot(is.na(cname) || is.null(cname) || is_string(cname))
+  path <- match.arg(path, choices = c("/", "/docs"))
+  tr <- target_repo(github_get = TRUE)
+  if (!isTRUE(tr$can_push)) {
+    ui_stop("
+      You don't seem to have push access for {ui_value(tr$repo_spec)}, which \\
+      is required to turn on GitHub Pages.")
+  }
+
+  gh <- function(endpoint, ...) {
```

Comment:
Given that you've used it 3 times, feels like should be it's own function?

Reply:
It has to be defined internally this way, because we're referencing `tr`. But we use this pattern so much that, after this PR, I am going to do something to make it a bit nicer everywhere.